src/librustc_typeck/check/cast.rs

```diff
@@ -111,17 +113,34 @@ enum CastError {
 }
 
 impl<'tcx> CastCheck<'tcx> {
-    pub fn new(expr: &'tcx hir::Expr, expr_ty: Ty<'tcx>, cast_ty: Ty<'tcx>, span: Span)
-               -> CastCheck<'tcx> {
-        CastCheck {
+    pub fn new<'a>(fcx: &FnCtxt<'a, 'tcx>,
+                   expr: &'tcx hir::Expr,
+                   expr_ty: Ty<'tcx>,
+                   cast_ty: Ty<'tcx>,
+                   cast_span: Span,
+                   span: Span)
+                   -> Result<CastCheck<'tcx>, ErrorReported> {
+        let check = CastCheck {
             expr: expr,
             expr_ty: expr_ty,
             cast_ty: cast_ty,
+            cast_span: cast_span,
             span: span,
+        };
+
+        // For better error messages, we try to check whether the
+        // target type is known to be sized now (we will also check
+        // later, once inference is more complete done).
+        if !fcx.type_is_known_to_be_sized(cast_ty, span) {
```

Comment:
@arielb1 

> What is constructor doing randomly emitting error messages?

As you well know, a constructor is just another function... seems like as good a time as any. And the return type (in particular, the `Result` with `ErrorReported`) clearly indicates that an error is reported. 

My thinking was basically that instantiating a cast check does some checks and returns, potentially, further checks that may need to be done later. I'm happy to rename.

> BTW, why do we need the early error message? It causes compilation errors in some cases.

Yes, this doesn't seem optimal, I agree. I was just preserving the existing behavior -- but it seems like it could go wrong, if only in unusual situations like `x as _` or `x as Bar<_>`

I originally moved this check to be done at the end, but then you get derived errors later. For example:

    let x = foo as SomeTrait;

will give an error because the type of `x` is `SomeTrait`, which is not `Sized`. (Actually, it then later gives a error about "cast to unsized type".)

Probably the best thing would be to rewrite `type_is_known_to_be_sized` to return "yes, no, maybe" and only error out on *no*. (Or, for simplicity, just check specifically for casts to traits or slices, and leave the more general function for the end.)